src/java.base/share/classes/java/time/InstantSource.java

```diff
@@ -0,0 +1,247 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package java.time;
+
+import java.time.Clock.FixedClock;
+import java.time.Clock.OffsetClock;
+import java.time.Clock.SourceClock;
+import java.time.Clock.SystemClock;
+import java.time.Clock.SystemInstantSource;
+import java.time.Clock.TickClock;
+import java.util.Objects;
+
+/**
+ * Provides access to the current instant.
+ * <p>
+ * Instances of this interface are used to find the current instant.
+ * As such, an {@code InstantSource} can be used instead of {@link System#currentTimeMillis()}.
+ * <p>
+ * The primary purpose of this abstraction is to allow alternate instant sources to be
+ * plugged in as and when required. Applications use an object to obtain the
+ * current time rather than a static method. This can simplify testing.
+ * <p>
+ * Best practice for applications is to pass an {@code InstantSource} into any method
+ * that requires the current instant. A dependency injection framework is one
+ * way to achieve this:
+ * <pre>
+ *  public class MyBean {
+ *    private InstantSource source;  // dependency inject
+ *    ...
+ *    public void process(Instant endInstant) {
+ *      if (source.instant().isAfter(endInstant) {
+ *        ...
+ *      }
+ *    }
+ *  }
+ * </pre>
+ * This approach allows an alternate source, such as {@link #fixed(Instant) fixed}
+ * or {@link #offset(InstantSource, Duration) offset} to be used during testing.
+ * <p>
+ * The {@code system} factory method provide a source based on the best available
+ * system clock This may use {@link System#currentTimeMillis()}, or a higher
+ * resolution clock if one is available.
+ *
+ * @implSpec
+ * This interface must be implemented with care to ensure other classes operate correctly.
+ * All implementations that can be instantiated must be final, immutable and thread-safe.
```

Comment:
(I'm not an openjdk reviewer)

I'm wondering if we can revisit this "immutable" requirement as it is being shuffled from `Clock` to `InstantSource`.

This precludes an implementation that might be useful for testing, where the time provided by a single `InstantSource` instance can be adjusted manually in the test code.  To me that is a notable downside (worth breaking the contract for), and it's not so clear what the upside is.  "Immutable" seems like an odd way to describe the system clock anyway.

Do you know if the people who already use their own `InstantSource` / `TimeSource` / `Supplier<Instant>` have a mutable version that they use in tests?  I would be surprised if they're all satisfied with `InstantSource.fixed`, but perhaps my intuition is wrong here.